core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -232,9 +244,21 @@ public void inboundDataReceived(ReadableBuffer frame, boolean endOfStream) {
      *
      * @param status the error status. Must not be {@link Status#OK}.
      */
-    public final void transportReportStatus(Status status) {
+    public final void transportReportStatus(final Status status) {
       Preconditions.checkArgument(!status.isOk(), "status must not be OK");
-      closeListener(status);
+      if (deframerClosed) {
+        deframerClosedTask = null;
+        closeListener(status);
+      } else {
+        deframerClosedTask =
+            new Runnable() {
+              @Override
+              public void run() {
+                closeListener(status);
+              }
+            };
+        closeDeframer(false);
```

Comment:
Previously the code notified the application immediately, right? Why isn't this `true`?